src/java/org/apache/cassandra/service/DataResolver.java

```diff
@@ -596,14 +596,16 @@ else if (DROP_OVERSIZED_READ_REPAIR_MUTATIONS)
     }
 
     private UnfilteredPartitionIterator extendWithShortReadProtection(UnfilteredPartitionIterator partitions,
-                                                                      InetAddress source,
-                                                                      DataLimits.Counter mergedResultCounter)
+                                                                      ResolveContext context,
+                                                                      int i)
     {
         DataLimits.Counter singleResultCounter =
             command.limits().newCounter(command.nowInSec(), false, command.selectsFullPartition(), enforceStrictLiveness).onlyCount();
 
-        ShortReadPartitionsProtection protection =
-            new ShortReadPartitionsProtection(source, singleResultCounter, mergedResultCounter);
+        ShortReadPartitionsProtection protection = new ShortReadPartitionsProtection(context.sources[i],
+                                                                                     () -> responses.clearUnsafe(i),
```

Comment:
@adelapena So this clears the responses only from the relevant source, and when the entire payload is consumed. WDYT? (Another benefit of doing it this way is that I think it will apply cleanly to `trunk`, where the short read protection classes have been moved out of `DataResolver`.)

Reply:
I think it's a nice improvement.